neps/nep-0455.md

```diff
@@ -0,0 +1,191 @@
+With compute costs, we will be able to roll out the read-only flat storage with minimal impact on deployed contracts.
+
+### Adjust the gas chunk limit
+We could continuously measure the chunk production time in nearcore clients and compare it to the gas burnt.
+If the chunk producer observes undercharging, it decreases the limit.
+If there is overcharging, the limit can be increased up to a limit of at most 1000 Tgas.
+
+Pros:
+- Prevents moderate undercharging from stalling the network
+- No protocol change necessary (as this feature is already [a part of the protocol](https://nomicon.io/Economics/Economic#transaction-fees)), we could easily experiment and revert if it does not work well
+
+Cons:
+- Very broad granularity --- undercharging in one parameter affects all users, even those that never use the undercharged parts
+- Dependence on validator hardware --- someone running overspecced hardware will continuously want to increase the limit, others might run with underspecced hardware and continuously want to decrease the limit
+- Malicious undercharging attacks are unlikely to be prevented by this --- a single 10x undercharged receipt still needs to be processed using the old limit.
+Adjusting 0.1% per block means 100 chunks can only change by a maximum of 1.1x and 1000 chunks could change up to x2.7
+- Conflicts with transaction and receipt limit --- A transaction or receipt can (today) use up to 300Tgas.
+The effective limit per chunk is `gas_limit` + 300Tgas since receipts are added to a chunk until one exceeds the limit and the last receipt is not removed.
+Thus a gas limit of 0gas only reduces the effective limit from 1300Tgas to 300Tgas, which means undercharging of > 4.33x is not covered.
+
+### Allow skipping chunks in the chain
+Slow chunk production can pose a liveness risk as the nodes expect a regular chunk production during normal operation.
+If processing the chunk takes more than expected, the block at the corresponding height will not be constructed in time and other nodes will start skipping this height effectively stalling the block production.
+
+We could extend the protocol to produce empty chunks for some of the shards within the block (effectively skipping them) when processing the chunk takes longer than expected.
+This way will still ensure a regular block production, at a cost of lower throughput of the network in that shard.
+The chunk should still be included in a later block to avoid stalling the affected shard.
+
+Pros:
+- Fast and automatic adaptation to the blockchain workload
+
+Cons:
+- Needs a voting mechanism to decide that processing of the chunk is slow (as opposed to the chunk producer being slow) and the chunk should be skipped within a block
```

Comment:
would this also have an impact on how the light client work, and the costs of state transition validations?

Reply:
To be honest, I don't know because this idea is not fully thought though.

I would defer to @abacabadabacaba to answer this, given that this idea is my sketch of a sketch proposed by @abacabadabacaba during one of the discussions that I've included for completeness.

Reply:
This solution is transparent to light clients. In the simplest version, it doesn't even require a protocol change, because chunks can be skipped already, for example, if the chunk producer is offline. Without a protocol change, however, chunk producers will be penalised for skipping chunks, so it may be desirable to distinguish between slow chunks and slow chunk producers, which would require some sort of voting. Or just consider missing chunks to be normal and not penalise chunk producers as much.
@akashin I suggest to add an explanation why voting (and a protocol change) is necessary with this solution.

Reply:
Makes sense to me, I've rephrased the point about on-chain voting mentioning the actual problem of distinguishing slow production from offline/malicious nodes